modules/cuml/src/umap.ts

```diff
@@ -0,0 +1,433 @@
+// Copyright (c) 2021, NVIDIA CORPORATION.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {MemoryData} from '@nvidia/cuda';
+import {DataFrame, Float32, Float64, Int64, Integral, Series} from '@rapidsai/cudf';
+import {DeviceBuffer} from '@rapidsai/rmm';
+import {compareTypes} from 'apache-arrow/visitor/typecomparator';
+
+import {CUMLLogLevels, MetricType} from './mappings';
+import {UMAPBase, UMAPInterface, UMAPParams} from './umap_base';
+import {dataframeToSeries, seriesToDataframe} from './utilities/array_utils';
+
+export type Numeric = Integral|Float32;
+
+export type outputType = 'dataframe'|'series'|'devicebuffer';
+
+export type returnTypeMap<T extends outputType, R extends Numeric> = {
+  'dataframe': DataFrame<{[P in number]: R}>,
+  'series': Series<R>,
+  'devicebuffer': DeviceBuffer
+}[T];
+
+export class UMAP<O extends outputType = any> {
+  public _umap: UMAPInterface;
+  public _embeddings: MemoryData|DeviceBuffer;
+  public outputType: O;
+
+  /**
+   * Initialize a UMAP object
+   * @param input: UMAPParams
+   * @param outputType: 'dataframe'|'series'|'devicebuffer'
+   *
+   * @example:
+   * ```typescript
+   * import {UMAP} from 'rapidsai/cuml';
+   *
+   * const umap = new UMAP({nComponents:2}, 'dataframe');
+   * ```
+   */
+  constructor(input: UMAPParams, outputType: O);
+  constructor(input: UMAPParams, outputType = 'dataframe') {
+    this._umap       = new UMAPBase(input);
+    this._embeddings = new DeviceBuffer();
+    this.outputType  = outputType as O;
+  }
+
+  protected _generate_embeddings(nSamples: number, dtype: Numeric): MemoryData {
+    return Series.sequence({type: dtype, size: nSamples * this._umap.nComponents, init: 0, step: 0})
+      .data.buffer;
+  }
+
+  // throw runtime error if type if float64
+  protected _check_type(X: Numeric) {
+    if (compareTypes(X, new Float64)) {
+      throw new Error('Expected input to be of type in [Integral, Float32] but got Float64');
+    }
+  }
```

Comment:
This is only checking if the type is `FLOAT64`, but will pass if the type is anything else (Utf8String, List, Struct, etc.). 

We should be more restrictive about the types we allow:

```suggestion
  // throw runtime error if type if not Bool8, Integral, or Float32
  protected _check_type<T extends Numeric>(X: T) {
    if (!allowedTypes.some((type) => compareTypes(X, type))) {
      throw new Error(
        `Expected input to be of type in [Integral, Float32] but got ${X.toString()}`);
    }
  }
```

Reply:
will `allowedTypes` be an explicit list like `[new Float32, new Int32, new Int8, new Bool8 ...]`?